### PR TITLE
redeem with private key

### PR DESCRIPTION
**Pull request by jellegerbrandy**
This PR implements the functionality from issue https://github.com/daostack/alchemy/issues/1036

### CODE REVIEW

File: src/components/Scheme/ReputationFromToken.tsx [hunk 1]
```diff
@@ -58,23 +62,43 @@ class ReputationFromToken extends React.Component<IProps, IState> {
...
+    const pk = queryValues["pk"] as string;
+    if (pk) {
+      console.log(`Found private key ${pk}`);
+      const arc = getArc();
+      try {
+        redeemerAddress = arc.web3.eth.accounts.privateKeyToAccount(pk).address;
+      } catch(err) {
+        throw Error(`Invalide private key: ${pk}`);
+      }
+      // redeemerAddress = "0x90F8bf6A479f320ead074411a4B0e7944Ea8c9C1";
```
Comment (dkent600) [inline src/components/Scheme/ReputationFromToken.tsx]: remove?

File: src/components/Scheme/ReputationFromToken.tsx [hunk 2]
```diff
@@ -116,13 +144,15 @@ class ReputationFromToken extends React.Component<IProps, IState> {
     if (alreadyRedeemed) {
```
Comment (A-Zak) [inline src/components/Scheme/ReputationFromToken.tsx]: Already redeemed should be based on the `redeemerAddress`, not the `accountAddress`
Looks like some nasty copy-pasting from line  101, probably worth to refactor out to one method ^_^

Comment (jellegerbrandy) [reply src/components/Scheme/ReputationFromToken.tsx]: good call, this was already a bug in the previous version..

File: src/actions/arcActions.ts [hunk 3]
```diff
@@ -135,15 +135,38 @@ export function redeemProposal(daoAvatarAddress: string, proposalId: string, acc
...
+      const gas = block.gasLimit - 100000;
+      const redeemMethod = contract.methods.redeem(addressToRedeem);
+      redeemMethod.gas = gas;
+      const signedTransaction = await arc.web3.eth.accounts.signTransaction(redeemMethod, privateKey);
+      dispatch(showNotification(NotificationStatus.Success, "Sending redeem transaction, please wait for it to be mined"));
+      console.log("Sending transaction");
+      try {
+        const receipt = await arc.web3.eth.sendSignedTransaction(signedTransaction.rawTransaction);
+        console.log(`Transaction sent: ${receipt}`);
+        dispatch(showNotification(NotificationStatus.Success, "Transaction was succesfull!"));
```
Comment (tibetsprague) [inline src/actions/arcActions.ts]: successful

File: src/actions/arcActions.ts [hunk 4]
```diff
@@ -135,15 +135,38 @@ export function redeemProposal(daoAvatarAddress: string, proposalId: string, acc
...
 
-export function redeemReputationFromToken(scheme: Scheme, addressToRedeem: string) {
+export function redeemReputationFromToken(scheme: Scheme, addressToRedeem: string, redeemerAddress: Address, privateKey: string|undefined) {
   return async (dispatch: Redux.Dispatch<any, any>) => {
-    const observer = operationNotifierObserver(dispatch, "Redeem reputation");
-    const reputationFromTokenScheme = scheme.ReputationFromToken as ReputationFromTokenScheme;
+    const arc = getArc();
+    if (privateKey) {
+      console.log("SIGNING WITH PK");
+      const reputationFromTokenScheme = scheme.ReputationFromToken as ReputationFromTokenScheme;
```
Comment (tibetsprague) [inline src/actions/arcActions.ts]: shouldnt this all be in the client library?

Comment (jellegerbrandy) [reply src/actions/arcActions.ts]: well, yes, you are probably right - it started as a one-liner but then I needed more. Also it needs to be deployed quickly, so I chose the quick solution.

File: src/actions/arcActions.ts [hunk 5]
```diff
@@ -135,15 +135,36 @@ export function redeemProposal(daoAvatarAddress: string, proposalId: string, acc
   };
 }
 
-export function redeemReputationFromToken(scheme: Scheme, addressToRedeem: string) {
+export function redeemReputationFromToken(scheme: Scheme, addressToRedeem: string, redeemerAddress: Address, privateKey: string|undefined) {
```
Comment (dkent600) [inline src/actions/arcActions.ts]: `redeemerAddress` is never referenced

Comment (dkent600) [reply src/actions/arcActions.ts]: i'd like to know why lint didn't catch this

Comment (jellegerbrandy) [reply src/actions/arcActions.ts]: good point

**Pull request merged**